fieldservice_recurring/models/fsm_recurring.py

```diff
@@ -0,0 +1,198 @@
+            else:
+                next_date = datetime.now()
+                interval = rec.fsm_frequency_id.interval
+                last_order = self.env['fsm.order'].search([
+                    ('fsm_recurring_id', '=', rec.id),
+                    ('stage_id', '!=', self.env.ref(
+                        'fieldservice.fsm_stage_cancelled').id)
+                ], offset=0, limit=1, order='scheduled_date_start desc')
+                if last_order:
+                    last_date = fields.Datetime.from_string(
+                        last_order.scheduled_date_start)
+                    next_date = last_date + relativedelta(days=+interval)
+                rec.next_schedule_date = next_date
+
+    @api.onchange('fsm_recurring_template_id')
+    def onchange_recurring_template_id(self):
+        if not self.fsm_recurring_template_id:
+            return
+        values = self.populate_from_template()
+        self.update(values)
+
+    def populate_from_template(self, template=False):
+        if not template:
+            template = self.fsm_recurring_template_id
+        vals = {
+            'fsm_frequency_id': template.fsm_frequency_id,
+            'description': template.description,
+            'fsm_order_template_id': template.fsm_order_template_id,
+            'company_id': template.company_id,
+        }
+        return vals
+
+    @api.model
+    def create(self, vals):
+        if vals.get('name', _('New')) == _('New'):
+            vals['name'] = self.env['ir.sequence'].next_by_code(
+                'fsm.recurring') or _('New')
+        return super(FSMRecurringOrder, self).create(vals)
+
+    @api.multi
+    def action_start(self):
+        for rec in self:
+            if not rec.start_date:
+                rec.start_date = datetime.now()
+            rec._create_next_order()
+            rec.write({'state': 'progress'})
+
+    @api.multi
+    def action_renew(self):
+        return self.action_start()
+
+    @api.multi
+    def action_cancel(self):
+        for order in self.fsm_order_ids.filtered(
+            lambda o: o.stage_id.is_closed is False
+        ):
+            order.action_cancel()
+        return self.write({'state': 'cancel'})
+
+    def _prepare_order_values(self):
+        schedule_date = self.next_schedule_date or self.start_date
+        days_early = self.fsm_frequency_id.buffer_early
+        days_late = self.fsm_frequency_id.buffer_late
+        earliest_date = fields.Datetime.from_string(schedule_date) + \
+            relativedelta(days=-days_early)
+        latest_date = fields.Datetime.from_string(schedule_date) + \
+            relativedelta(days=+days_late)
+        return {
+            'fsm_recurring_id': self.id,
+            'customer_id': self.customer_id.id,
+            'location_id': self.location_id.id,
+            'scheduled_date_start': schedule_date,
+            'request_early': str(earliest_date),
+            'request_late': str(latest_date),
```

Comment:
This isn't setting the value on the request late field as expected because it is a computed field of fsm.order